### PR TITLE
Allow eta-expansion of methods with dependent types

### LINKED ISSUE #12641: `StringBuilder` change in 2.13.9 (`append` returning `this.type`) exposes poor interaction between eta-expansion and overloading
## Reproduction steps

Scala version: 2.13.9

```scala
Welcome to Scala 2.13.9 (OpenJDK 64-Bit Server VM, Java 1.8.0_322).
Type in expressions for evaluation. Or try :help.

scala> :paste
// Entering paste mode (ctrl-D to finish)

case class TestObj(testValue: Option[String] = None)
def appendToBuilder(obj: TestObj, builder: StringBuilder): Unit = {
  obj.testValue.foreach(builder.append)
}


// Exiting paste mode, now interpreting.


  obj.testValue.foreach(builder.append)
                                ^
<pastie>:3: error: type mismatch;
 found   : (x: Char): builder.type <and> (x: Double): builder.type <and> (x: Float): builder.type <and> (x: Long): builder.type <and> (x: Int): builder.type <and> (x: Short): builder.type <and> (x: Byte): builder.type <and> (x: Boolean): builder.type <and> (s: StringBuilder): builder.type <and> (cs: CharSequence): builder.type <and> (s: String): builder.type <and> (x: Any): builder.type (with underlying type (x: Char): builder.type <and> (x: Double): builder.type <and> (x: Float): builder.type <and> (x: Long): builder.type <and> (x: Int): builder.type <and> (x: Short): builder.type <and> (x: Byte): builder.type <and> (x: Boolean): builder.type <and> (s: StringBuilder): builder.type <and> (cs: CharSequence): builder.type <and> (s: String): builder.type <and> (x: Any): builder.type)
 required: String => ?
```

This code can be compiled with Scala 2.13.8 

```scala
Welcome to Scala 2.13.8 (OpenJDK 64-Bit Server VM, Java 1.8.0_322).
Type in expressions for evaluation. Or try :help.

scala> :paste
// Entering paste mode (ctrl-D to finish)

case class TestObj(testValue: Option[String] = None)
def appendToBuilder(obj: TestObj, builder: StringBuilder): Unit = {
  obj.testValue.foreach(builder.append)
}


// Exiting paste mode, now interpreting.

class TestObj
def appendToBuilder(obj: TestObj, builder: StringBuilder): Unit
```

## Problem
The same code can be compiled with 2.13.8, but fails with 2.13.9

### PR DESCRIPTION
**Pull request by lrytz**
Fixes https://github.com/scala/bug/issues/12641

### CODE REVIEW

File: test/files/run/eta-dependent.scala [hunk 1]
```diff
@@ -0,0 +1,59 @@
...
+      val f2: String => a2.type = a2.foo
+      val f3: String => A = a2.foo
+    }
+  }
+}
+
+import scala.tools.partest.ReplTest
+
+object Test extends ReplTest {
+  override protected def normalize(s: String): String = s.replaceAll("(Lambda).*", "$1")
```
Comment (som-snytt) [inline test/files/run/eta-dependent.scala]: there is `ReplTest with Lambdaless`

File: src/compiler/scala/tools/nsc/typechecker/Typers.scala [hunk 2]
```diff
@@ -3202,12 +3202,9 @@ trait Typers extends Adaptations with Tags with TypersTracking with PatternTyper
...
 
-      if (tree.tpe.isDependentMethodType) DependentMethodTpeConversionToFunctionError(tree, tree.tpe) // TODO: support this
-      else {
-        val expansion = etaExpand(tree, context.owner)
-        if (context.undetparams.isEmpty) typed(expansion, mode, pt)
-        else instantiate(typed(expansion, mode), mode, pt)
-      }
+      val expansion = etaExpand(tree, context.owner)
+      if (context.undetparams.isEmpty) typed(expansion, mode, pt)
+      else instantiate(typed(expansion, mode), mode, pt)
```
Comment (som-snytt) [inline src/compiler/scala/tools/nsc/typechecker/Typers.scala:3207]: I don't understand the restriction yet, but `etaExpand` still has adriaanm comment that not supported.

Comment (lrytz) [reply src/compiler/scala/tools/nsc/typechecker/Typers.scala:3207]: Thanks! I'll be off for a week 🇷🇴 and don't have more time right now to try to break it, will get back to it after.

Comment (som-snytt) [reply src/compiler/scala/tools/nsc/typechecker/Typers.scala:3207]: Enjoy! Insert joke about children as dependent types.

Comment (lrytz) [reply src/compiler/scala/tools/nsc/typechecker/Typers.scala:3207]: Added a few more tests, AFAICT `etaExpand` works well with dependent method types

**Comment by SethTisue** [comment]
@som-snytt care to give this a second round of review?

**Review by som-snytt — APPROVED** [review]
Lukas could have just told me in September how badly I needed this.

**Pull request merged**